src/main/comments.js

```diff
@@ -387,9 +389,30 @@ function printLeadingComment(commentPath, print, options) {
   // Leading block comments should see if they need to stay on the
   // same line or not.
   if (isBlock) {
+    const commentsHasNewline = hasNewline(
+      options.originalText,
+      options.locEnd(comment)
+    );
+    const grandParent = commentPath.getParentNode(1);
+    const leftKey =
+      grandParent &&
+      (grandParent.type === "VariableDeclarator"
```

Comment:
Sorry, I really can't review this PR properly at the moment, but this is not right. The files in `src/main` should not mention language-specific things like AST node types. Language-specific code like this should reside in `language-js`. For example, check how I added `getGapRegex`.